### ISSUE TITLE
Memory stream commands: block data read and writes

### ISSUE THREAD

**Issue by tomxp411** [X16Community/x16-rom#44]
These commands are intended to speed up reading and writing strings and numeric data to and from banked memory.  These will be useful for manipulating directory data in RAM. 

Use banked RAM as a RAM disk. SAVE, OPEN, GET#, and INPUT# will read and write directly to banked RAM. 

`OPEN 1,1,1,"$2000"` or `OPEN 1,1,$2000` Opens a stream starting at the offset specified in the filename (or SA). Ideally, this value is an _index into banked RAM_, not a fixed address. 

So a value of 0 or blank would start at bank 1, address $A000. 8192 would be the start of bank 1, and so on. The goal is to treat banked memory as much like a file stream as possible, so we should abstract away the addressing details of banked memory.

`PRINT#` would write directly to memory at the current address pointer and advances the pointer by the number of bytes in the string, including an optional trailing CR.

`GET#` reads from memory at the current address pointer and advances the pointer by 1 byte.

`GET#1,A$,<length>` would read _length_ bytes from the data stream and advances the pointer by _length_ bytes. 

`INPUT#1,A$` reads a line of text into A$ and advances the pointer to the character after: comma (,), null (zero), or CR (13), as well as any other BASIC field delimiters. 

Also, we have inconsistent behavior when INPUT# encounters a null (zero) byte. Ideally, the null byte should be treated as an "end of line", and the read pointer should be set to the byte immediately following the null.

**Comment by tomxp411** [comment]
As an example, reading a directory in BASIC becomes very simple: 

100 BLOAD "$",1,$A000
110 STREAM $A000
120 STARTADDR=SGETI(2)
130 PTR=SGETI(2) : IF PTR = 0 THEN 200
140 PRINT SGETI(2) : REM FILE LENGTH
150 PRINT SREAD$(-255) : REM FILENAME
160 GOTO 130 
200 END

**Comment by mooinglemur** [comment]
After sleeping on it, I'm leaning toward a new function or statement to fetch a C-style string from file I/O and leaving `INPUT#` alone.  There's a lot of legacy baggage in there like field separators that seem like it'd really get in the way.

I think a function makes more sense since it always returns a single string, but then it loses its `xxxx#` consistency for file I/O
`A$=GETCSTR$(8)`

any thoughts?

**Comment by mooinglemur** [comment]
Perhaps `GETB$()` vs `GETCS$()`

**Comment by tomxp411** [comment]
That works for me. You’ll probably get some people confused by what GETB$ means, maybe GETFL$. I honestly can’t think of a better, short name, myself…. So I guess go with what feels right.

**Comment by mooinglemur** [comment]
The memory as abstraction layer would be quite a bit more complex than I originally thought.  There's just not a good way to fit things in with how things are currently implemented.

If we were to do these sorts of enhancements, I think it might be best to implement the file I/O read commands and or/the memory stream commands, such as `SINPUT#1,A$` and `BINPUT#1,A$,4` for string input and binary input respectively.
For memory commands, A$ = SPEEK$($A000) for C string peek from memory, A$ = BPEEK$($A000,4) for a multibyte peek into a string variable.

@tomxp411 can you rank these four potential operations in the order that you'd most want to see?  I'm thinking the file I/O ones would be the most useful, but I don't spend a ton of time in BASIC so I don't know which are the worst pain points.

**Comment by tomxp411** [comment]
It's a flip of the coin. I can read a large file into banked RAM and work with it using SPEEK$, or I can use SINPUT#1,A$ to read lines from a file. 

If I had to guess which I'm most likely to use _right now_, it's probably the file I/O stuff. If it gets to the point where I need to load arbitrary strings from structured data in banked RAM, I've already started designing a small ML subroutine library to do that...
